### PR TITLE
Limit the number of spawned go-rountines in RetrieveBlockData

### PR DESCRIPTION
**Pull request by jclyons52**
#357 

## Description
Set a limit of 100 go routines
used solution from this article: https://medium.com/happytech/limiting-goroutines-with-semaphore-b8ccc248e5f0

Closes: #357

### CODE REVIEW

File: p2p/ipld/read.go
Comment (jclyons52) [inline p2p/ipld/read.go:49]: I'm new to go so not sure if this is needed but passing in all arguments to resolve a warning https://stackoverflow.com/questions/40326723/go-vet-range-variable-captured-by-func-literal-when-using-go-routine-inside-of-f

File: p2p/ipld/read_test.go
Comment (Wondertan) [inline p2p/ipld/read_test.go:139]: 👍🏻

File: p2p/ipld/read.go
Comment (Wondertan) [inline p2p/ipld/read.go:48]: Pls, pass context from the first func argument.

File: p2p/ipld/read.go
Comment (Wondertan) [inline p2p/ipld/read.go:34]: * Pls move the magic number to the const
* I guess the number you've chosen is simply a guess. To determine the ideal number we would need some additional analysis, but my intuition says that `100` is not enough and something like `512` would play well for us here. Althouh, I would put something like '1024' instead, but I concerned about race detector erroring with such figure, as tests may run multiple `RetreieveBlockData`. Can you play for a bit with a higher number?

File: p2p/ipld/read.go
Comment (Wondertan) [reply p2p/ipld/read.go:49]: This is needed to ensure goroutines does not read from vars that gets rewritten on every loop iteration. In this case, you need to pass through: `rootCid`, `row` and `col`. `dag` and `isRow` won't change on loop iterations, so you can pass them directly to `retrieveShare`, and have -2 vars to the goroutine.

**Review by Wondertan — CHANGES_REQUESTED** [review]
Hey, @jclyons52. Thanks for your contribution! That's an interesting choice for the first PR, indeed.

The end goal for this change would be removing the special skip case for tests, what you already did 👍🏻; and ensuring tests pass with no race detector errors.

**Comment by jclyons52** [comment]
Thanks for the review @Wondertan! I've pushed up the requested changes but I'm still playing around with different limits. Hold off on merging as I'm also investigating an issue using the parent context: `panic: semaphore: released more than held`

**Comment by jclyons52** [comment]
@Wondertan @evan-forbes I realised this morning that it was actually the return channels that were keeping the go-routines alive. Without the limits on those I'm actually not even able to trigger the race detector, do they need to have that size limit? if not should I remove the semaphore? In theory it could still help but in practice it seems not to matter.

**Comment by jclyons52** [comment]
ok so I've updated the loop to run in a goroutine, this allows sc.wait to start getting results from the output channel before the for loops have finished pushing them in. Previously the output channel had a limit of 1 so all the go routines would block until sc.wait started pulling messages after the the loop had finished, so at the time of the loop finishing there would be as many go routines as there were loop iterations, the semaphore could not resolve this as it had to release before the message was sent to the channel in order not to deadlock. On my local machine I got optimal performance with ~30 max go routines but that's probably because I've only got 2 cpu cores, optimal number will depend on the sort of infrastructure it's running on I suppose. Might be worth setting the max number based on CPU cores.

**Comment by jclyons52** [comment]
ok so one last issue, on my local I'm running it with `ctx, cancel := context.WithCancel(ctx)` rather than `ctx, cancel := context.WithTimeout(ctx, time.Second*2)` because it was timing out when I was debugging. With that set back to the timeout context I'm getting the race condition again, seems that I'm not handling timeouts correctly with the semaphore.

**Comment by evan-forbes** [comment]
Providing more context for this PR, the original reason that we wanted to be able to optionally limit the number of goroutines was so that we could test the max square size while also using the `-race` flag. When we're not testing, we will likely care a lot less as to the number of goroutines used. 

Even the though the current goroutine limit is set 512 in this PR, when we call `dag.Get` in `GetLeafData`, that function spins up **a lot** of goroutines. If we have the ability to limit goroutines, then at least while we're testing, we can just limit the number of spun up possibly as low as 50, while also increasing the timeout. This would allowing us to use the `-race` flag while also testing for the max square size. 

The problem that I'm encountering while playing around with this awesome change, is that any time a cancel is sent via the context, whether is be a timeout or otherwise, the semaphore panics. 
```
panic: semaphore: released more than held
```
This is a issue, as we don't want to panic upon reaching a timeout. We still might be able to use this semaphore code, but if this keeps giving us trouble, fortunately there are a ton of different ways to limit the number goroutines used. One of the most versatile templates that I've found is the [goroutine worker pool](https://gobyexample.com/worker-pools). 

If we do use the semaphore code, we might need an additional test to ensure that it does not panic when canceled.

File: p2p/ipld/read.go
Comment (liamsi) [inline p2p/ipld/read.go:61]: I feel there are ways to achieve the same more idiomatically (without using x/sync/semaphore).

**Review by liamsi — COMMENTED** [review]
Thanks a lot @jclyons52 🙏🏼 

I agree with @evan-forbes' assessment. Just wanted to add a few inline comments.

File: p2p/ipld/read.go
Comment (liamsi) [reply p2p/ipld/read.go:21]: I see, thanks. That sentence is more a general remark. I think in general we should try to use the bandwidth we got.

> we get more routines just waiting for Bitswap walking down the tree

Just for my understanding: but wont the several routines trigger several bitswap requests in parallel, which can fill out more bandwidth? Sure due to the tree structure, you have to walk down top to bottom anyways but won't some of the networking IO still happen in parallel? e.g. when the first two children of the root are retrieved you can think of it as two threads (for the left and the right half).
But yeah spawning routines in way that leverages knowledge about bitswap to a) minimize the number of spawned routines while b) maximizing the use of available bandwidth, sounds like a good idea (and then we can likely do with even less then 512 routines).

Comment (Wondertan) [reply p2p/ipld/read.go:21]: @liamsi, I was rewriting the message constantly, till I am not satisfied with the explanation. Pls, revisit it

Comment (liamsi) [reply p2p/ipld/read.go:21]: Yeah, makes sense to me. Thanks!

Comment (Wondertan) [reply p2p/ipld/read.go:21]: > but wont the several routines trigger several bitswap requests in parallel

No, we don't know what to request initially to request in parallel as we work with trees here. Traversing them causing roundtripping, we don't need lots of routines to initiate new round trip.

Comment (Wondertan) [reply p2p/ipld/read.go:21]: >  but won't some of the networking IO still happen in parallel?

It happens in parallel, but the parallelism is strictly controlled by Bitswap

Comment (Wondertan) [reply p2p/ipld/read.go:21]: The next step would be to run one routine per DARoot. As an optimization, it is also possible to run one routine just per every Row or Col roots and we would still get whole data, but without  inner nodes of Row or Col roots.

Comment (Wondertan) [reply p2p/ipld/read.go:21]: > It happens in parallel, but the parallelism is strictly controlled by Bitswap

Thinking about DAG with Bitswap is different from HTTP. In the latter, you can spin up more routines with requests to use more bandwidth, as it straightly sends it to remote server, in former case, spawning more routines does not affect anything.

**Review by Wondertan — CHANGES_REQUESTED** [review]
@jclyons52, thanks again for tackling this. The PR initiated more thinking and discussion over the problem, resulting in #429. But anyway, we would be happy to merge this PR, as it specifically fixes #357. After, we encourage you to go to the next level 🆙  and try #429.

However, before merging this, we need to ensure:
* CI is not failing and tests are passing
* Code styling/correctness issues are fixed(like checking for in Acquire)
* No unexpected panic happens

File: p2p/ipld/read_test.go
Comment (jclyons52) [reply p2p/ipld/read_test.go:166]: I think this might just have been an issue with my crappy mac. Running the test normally works fine but the -race flag slows it down to about 45 seconds on my 2 cpu core mac. I'll try it out on my PC and see how I go, probably a better machine for multithreading applications.

Comment (jclyons52) [reply p2p/ipld/read_test.go:166]: 512 is just the number of iterations of that loop that can run simultaneously but if dag.Get is spinning up a high number of go routines the total could be significantly higher.

File: p2p/ipld/read.go
Comment (evan-forbes) [inline p2p/ipld/read.go:60]: I think we will want to `defer close(sem)` in the beginning of this goroutine, and then return if the context is finished on line 49, just to wrap up loose ends in the case of a prematurely canceled context

**Comment by liamsi** [comment]
@jclyons52, we decided that a lot of the code in this PR is going to be moved into a separate repo (very soon). It probably makes sense to close this PR for now unless you want to finish it before that.

**Comment by liamsi** [comment]
Closing this for now. Thanks again for your contribution @jclyons52! This code will be refactored and moved into https://github.com/celestiaorg/celestia-node

**Pull request closed without merging**
